api.go

```diff
@@ -0,0 +1,181 @@
+package tradingview
+
+import (
+	"encoding/json"
+	"errors"
+	"strconv"
+	"strings"
+
+	"github.com/gorilla/websocket"
+)
+
+/*
+Switches the timezone the data is viewed on for the current session
+*/
+func (api *API) SwitchTimezone(timezone string) error {
+	return api.sendWriteThread("switch_timezone", append([]interface{}{api.session.chart.key}, timezone))
+}
+
+/*
+Sends the data to the server in the format it requests
+*/
+func (api *API) sendServerMessage(name string, args []interface{}) error {
+	// change input into requested format
+	message, err := json.Marshal(
+		map[string]interface{}{
+			"m": name,
+			"p": args,
+		},
+	)
+
+	// ensure the data was created without issues
+	if err != nil {
+		return err
+	}
+
+	// send the message to the server
+	return api.ws.WriteMessage(websocket.TextMessage, []byte(SEPARATOR+strconv.Itoa(len(message))+SEPARATOR+string(message)))
+}
+
+/*
+Parse the message from the server
+*/
+func (api *API) parseServerMessage(buffer string) error {
+	// separate each message by the separator
+	msgs := strings.Split(buffer, SEPARATOR)
+
+	// for each message
+	for _, msg := range msgs {
+		// parse the message through JSON
+		var res map[string]interface{}
+		err := json.Unmarshal([]byte(msg), &res)
+
+		// the message was not JSON
+		if err != nil {
+			// send the message back if it contains the identifier
+			if strings.Contains(msg, "~h~") {
+				if err = api.ws.WriteMessage(websocket.TextMessage, []byte(SEPARATOR+strconv.Itoa(len(msg))+SEPARATOR+msg)); err != nil {
+					return err
+				}
+			}
+
+			continue
+		}
+
+		// the message was valid JSON
+		// unlock the mutex if the requested string has been returned by the server
+		if api.halted.on != "" && res["m"] == api.halted.on {
+			api.halted.mutex.Unlock()
+			api.halted.on = "" // resetting the mutex
+
+			// otherwise handle the response
+		} else {
+			api.handler(res, msg)
```

Comment:
Errors are not handled ? 😢

Reply:
damn I was rushing that one ig, I'll check it out in a bit